crates/hir-def/src/item_tree.rs

```diff
@@ -115,20 +115,36 @@ impl ItemTree {
             return Default::default();
         }
 
-        let ctx = lower::Ctx::new(db, file_id);
         let mut top_attrs = None;
         let mut item_tree = match_ast! {
             match syntax {
                 ast::SourceFile(file) => {
+                    let ctx = lower::Ctx::new(db, file_id);
                     top_attrs = Some(RawAttrs::new(db, &file, ctx.hygiene()));
                     ctx.lower_module_items(&file)
                 },
                 ast::MacroItems(items) => {
+                    // If a fn-like macro is defined in a trait, it should have the inherited visibility
+                    let trait_node = {
+                        if let Some((_, original_node)) = file_id.original_call_node(db.upcast()) {
```

Comment:
This doesn't quite work with nested macro invocations, we should use `file_id.call_node` here and take the parent before searching(since the call node could be an attributed item), and then probably loop up `ast::MacroItems` until we find an item(that is not a macro-call) and check whether that is a trait or not.

Reply:
maybe we should handle the inherited visibility during collection instead?

Reply:
It might make more sense to in collection? I am not sure how we treat visibility in these parts of the code base

Reply:
Handling it in collection is intuitive. But how could we pass the inherited visibility to the `MacroFile` so that we can use it when lowering the `MacroFile`?

Reply:
Okay so the relevant code you should need to touch for this is here https://github.com/rust-lang/rust-analyzer/blob/master/crates/hir-def/src/data.rs#L460-L531

When collecting for a trait we just have to re-use the traits visibility there instead then. How to handle macros should be apparent from there as well I believe since that code also has to handle macros and attribute already.

Reply:
Thanks for the tips! Actually, I've noticed that the trait visibility is available there. I'm just wondering how to make it available when lowering the `MacroFile` for the item tree.

I was thinking to add a field `inherited_visibility: RawVisibility` to the `MacroFile`, however, `RawVisibility` is not accessible from the hir-expand crate.

I also tried to add `inherited_visibility: RawVisibilityId` as a query parameter to the `file_item_tree_query` function. But I didn't figure out where to intern the `RawVisbility`. Any suggestions?

Reply:
The idea would be that the `ItemTree` reflects what's literally written in the code, and the inherited visibility only comes into play at the collection stage, where we know that we're collecting trait methods.

Reply:
Thanks for the info! This is my 1st time to contribute to this project. I'd like to learn and try if you don't mind :) I will switch this PR to draft now.